### PR TITLE
Support dynamic names for valueFrom resources in composition

### PR DESCRIPTION
**Pull request by barney-s**
Add valuesFrom.resourceRef.nameSuffix to support the use case where he dependent resource name is dynamically generated by a previous stage.

For now it supports the pattern: <inputapi.name><suffix>
This is limiting but provides a way to achieve it.

### CODE REVIEW

File: experiments/compositions/composition/api/v1/composition_types.go [hunk 1]
```diff
@@ -32,10 +32,11 @@ type ResourceRef struct {
 	// resource: ServiceIdentity.serviceusage.cnrm.cloud.google.com/v1beta1//sqladmin.googleapis.com
 
 	// OPTION 2
-	Group   string `json:"group" protobuf:"bytes,1,name=group"`
-	Version string `json:"version,omitempty" protobuf:"bytes,2,opt,name=version"`
-	Kind    string `json:"kind" protobuf:"bytes,3,name=kind"`
-	Name    string `json:"name" protobuf:"bytes,4,name=name"`
+	Group      string `json:"group" protobuf:"bytes,1,name=group"`
```
Comment (justinsb) [inline experiments/compositions/composition/api/v1/composition_types.go:35]: Do we need the proto annotations?

File: experiments/compositions/composition/api/v1/composition_types.go [hunk 2]
```diff
@@ -153,9 +154,17 @@ func (s *Composition) Validate() bool {
 	s.Status.ClearCondition(ValidationFailed)
 	// Validate Expanders
 	message := ""
-	for i, expander := range s.Spec.Expanders {
+	for expanderIndex, expander := range s.Spec.Expanders {
 		if expander.Name == "" {
-			message += fmt.Sprintf("Expander: %d missing Name", i)
+			message += fmt.Sprintf("Expander: %d missing Name", expanderIndex)
```
Comment (justinsb) [inline experiments/compositions/composition/api/v1/composition_types.go:159]: Nit: something like `spec.expanders[%d] missing name` might be clearer.  We could even use the validation errors from apimachinery, but it'll end up with something like  `spec.expanders[%d]`  IIRC

**Comment by justinsb** [comment]
Some comments, but nothing blocking.  I'll approve & lgtm & hold, feel free to clear hold once you've seen comments.

/approve
/lgtm
/hold

File: experiments/compositions/composition/api/v1/composition_types.go (hunk 1 above)
Comment (barney-s) [reply experiments/compositions/composition/api/v1/composition_types.go:35]: may be not. I started off with @cheftako's initial code and continued having it. 
We can remove if not needed for k8s usecase.

Comment (barney-s) [reply experiments/compositions/composition/api/v1/composition_types.go:34]: Added a TODO in the file.

**Review by barney-s — COMMENTED** [review]
resolved comments

**Comment by zicongmei** [comment]
Adding `nameSuffix` is not going to solve the problem. There can be multiple resources with a same suffix.

**Pull request merged**